Quitting Helix when `ocamllsp` was active causes delay and exits with an error.

When quitting the editor from an OCaml source file with `ocamllsp` active, there is a noticeable delay before exiting with the following error:

```
Error: Timed out waiting for language servers to shutdown
```

This part of the log is relevant:

```
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "/-----------------------------------------------------------------------\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Internal error: Uncaught exception.\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Jsonrpc: json conversion failed: invalid structured value\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Raised at Jsonrpc__Import.Json.error in file \"jsonrpc/src/import.ml\", line 31, characters 23-50\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Called from Jsonrpc__Import.Option.map in file \"jsonrpc/src/import.ml\", line 7, characters 21-26\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Called from Jsonrpc.Packet.t_of_fields in file \"jsonrpc/src/jsonrpc.ml\", line 271, characters 21-77\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Called from Lsp__Io.Make.read in file \"lsp/src/io.ml\", line 108, characters 24-57\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Called from Fiber__Scheduler.exec in file \"src/fiber/scheduler.ml\", line 73, characters 8-11\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Re-raised at Stdune__Exn.raise_with_backtrace in file \"otherlibs/stdune/exn.ml\" (inlined), line 36, characters 27-56\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Called from Stdune__Exn_with_backtrace.reraise in file \"otherlibs/stdune/exn_with_backtrace.ml\", line 18, characters 33-71\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "| Called from Fiber__Scheduler.exec in file \"src/fiber/scheduler.ml\", line 73, characters 8-11\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "\\-----------------------------------------------------------------------\n"
2023-01-05T23:13:31.638 helix_lsp::transport [ERROR] err <- "\n"
```

That is stderr output from the language server. It looks like it can't parse the shutdown message. It will take some investigation into the ocamllsp codebase but this looks like an upstream bug to me.

--- a/helix-lsp/src/jsonrpc.rs
+++ b/helix-lsp/src/jsonrpc.rs
@@ -170,6 +170,10 @@ impl Params {
         serde_json::from_value(value)
             .map_err(|err| Error::invalid_params(format!("Invalid params: {}.", err)))
     }
+
+    pub fn is_none(&self) -> bool {
+        self == &Params::None
+    }
 }
 
 impl From<Params> for Value {
@@ -187,7 +191,7 @@ impl From<Params> for Value {
 pub struct MethodCall {
     pub jsonrpc: Option<Version>,
     pub method: String,
-    #[serde(default = "default_params")]
+    #[serde(default = "default_params", skip_serializing_if = "Params::is_none")]
     pub params: Params,
     pub id: Id,
 }
@@ -197,7 +201,7 @@ pub struct MethodCall {
 pub struct Notification {
     pub jsonrpc: Option<Version>,
     pub method: String,
-    #[serde(default = "default_params")]
+    #[serde(default = "default_params", skip_serializing_if = "Params::is_none")]
     pub params: Params,
 }
 
@@ -332,6 +336,33 @@ fn notification_serialize() {
         serialized,
         r#"{"jsonrpc":"2.0","method":"update","params":[1,2]}"#
     );
+}
+
+#[test]
+fn serialize_skip_none_params() {
+    use serde_json;
+
+    let m = MethodCall {
+        jsonrpc: Some(Version::V2),
+        method: "shutdown".to_owned(),
+        params: Params::None,
+        id: Id::Num(1),
+    };
+
+    let serialized = serde_json::to_string(&m).unwrap();
+    assert_eq!(
+        serialized,
+        r#"{"jsonrpc":"2.0","method":"shutdown","id":1}"#
+    );
+
+    let n = Notification {
+        jsonrpc: Some(Version::V2),
+        method: "exit".to_owned(),
+        params: Params::None,
+    };
+
+    let serialized = serde_json::to_string(&n).unwrap();
+    assert_eq!(serialized, r#"{"jsonrpc":"2.0","method":"exit"}"#);
 }
 
 #[test]

